core/Pos/Core/Genesis/Types.hs

```diff
@@ -96,13 +103,50 @@ instance Buildable GenesisWStakeholders where
     build (GenesisWStakeholders m) =
         bprint ("GenesisWStakeholders: "%mapJson) m
 
+-- | This type contains genesis state of heavyweight delegation. It
+-- wraps a map where keys are issuers (i. e. stakeholders who
+-- delegated) and values are proxy signing keys. There are some invariants:
+-- 1. In each pair delegate must differ from issuer, i. e. no revocations.
+-- 2. PSKs must be consistent with keys in the map, i. e. issuer's ID must be
+--    equal to the key in the map.
+-- 3. Delegates can't be issuers, i. e. transitive delegation is not supported.
+--    It's not needed in genesis, it can always be reduced.
+newtype GenesisDelegation = UnsafeGenesisDelegation
+    { unGenesisDelegation :: HashMap StakeholderId ProxySKHeavy
+    } deriving (Show, Eq)
+
+-- | Empty 'GenesisDelegation'.
+noGenesisDelegation :: GenesisDelegation
```

Comment:
I'd go with `deriving Monoid` instead.

Reply:
If I have `GenesisDelegation` where `a` delegates to `b` and `GenesisDelegation` where `b` delegates to `c`, their `mappend` will be illegal.